src/model/evaluation.jl

```diff
@@ -1,87 +1,204 @@
+    loglikelihood = 0.0
     evaluation_env = deepcopy(model.evaluation_env)
+
     for (i, vn) in enumerate(model.graph_evaluation_data.sorted_nodes)
         is_stochastic = model.graph_evaluation_data.is_stochastic_vals[i]
         is_observed = model.graph_evaluation_data.is_observed_vals[i]
         node_function = model.graph_evaluation_data.node_function_vals[i]
         loop_vars = model.graph_evaluation_data.loop_vars_vals[i]
-        if_sample = sample_all || !is_observed # also sample if not observed, only sample conditioned variables if sample_all is true
+        if_sample = sample_all || !is_observed
+
         if !is_stochastic
             value = node_function(evaluation_env, loop_vars)
             evaluation_env = setindex!!(evaluation_env, value, vn)
         else
             dist = node_function(evaluation_env, loop_vars)
             if if_sample
-                value = rand(rng, dist) # just sample from the prior
+                value = rand(rng, dist)
             else
                 value = AbstractPPL.get(evaluation_env, vn)
             end
-            if model.transformed
-                # see below for why we need to transform the value
+
+            if transformed
                 value_transformed = Bijectors.transform(Bijectors.bijector(dist), value)
-                logp +=
+                logp =
                     Distributions.logpdf(dist, value) + Bijectors.logabsdetjac(
                         Bijectors.inverse(Bijectors.bijector(dist)), value_transformed
                     )
             else
-                logp += Distributions.logpdf(dist, value)
+                logp = Distributions.logpdf(dist, value)
             end
+
+            if is_observed
+                loglikelihood += logp
+            else
+                logprior += logp
+            end
+
             evaluation_env = setindex!!(evaluation_env, value, vn)
         end
     end
-    return evaluation_env, logp
+
+    return evaluation_env,
+    (
+        logprior=logprior,
+        loglikelihood=loglikelihood,
+        tempered_logjoint=logprior + temperature * loglikelihood,
+    )
 end
 
-function AbstractPPL.evaluate!!(model::BUGSModel)
-    logp = 0.0
-    evaluation_env = deepcopy(model.evaluation_env)
+"""
+    function evaluate_with_env!!(
+        model::BUGSModel,
+        evaluation_env=deepcopy(model.evaluation_env);
+        temperature=1.0,
+        transformed=true,
+    )
+
+Evaluate model using current values in the evaluation environment.
+
+# Arguments
+- `model`: The BUGSModel to evaluate
+- `temperature`: Temperature for tempering the likelihood (default 1.0)
+- `transformed`: Whether to compute log density in transformed space (default true)
+
+# Returns
+- `evaluation_env`: Updated evaluation environment
+- `(logprior, loglikelihood, tempered_logjoint)`: NamedTuple of log densities
+"""
+function evaluate_with_env!!(
+    model::BUGSModel,
+    evaluation_env=deepcopy(model.evaluation_env);
```

Comment:
Are you sure you need a deepcopy? A lot of people use it when they want a "safe" copy, but it's often considered bad practice: [[1](https://github.com/JuliaLang/julia/issues/42796#issuecomment-951232853), [2](https://ericphanson.com/blog/2024/why-you-might-avoid-deepcopy-in-julia/)]

Reply:
This is one of things on my TODO list!

I traditionally use it for safety reasons, but I am moving away from it because with BUGS models, I actually know what is being mutated.